src/iceberg/util/logger.h

```diff
@@ -0,0 +1,285 @@
+
+  template <typename... Args>
+  void LogImpl(LogLevel level, std::string_view format_str, Args&&... args) const {
+    if constexpr (sizeof...(args) > 0) {
+      std::string formatted_message =
+          std::vformat(format_str, std::make_format_args(args...));
+      LogRawImpl(level, formatted_message);
+    } else {
+      LogRawImpl(level, std::string(format_str));
+    }
+  }
+
+  template <typename... Args>
+  void LogWithLocationImpl(LogLevel level, std::string_view format_str,
+                           const std::source_location& location, Args&&... args) const {
+    std::string message;
+    if constexpr (sizeof...(args) > 0) {
+      message = std::vformat(format_str, std::make_format_args(args...));
+    } else {
+      message = std::string(format_str);
+    }
+    LogWithLocationRawImpl(level, message, location);
+  }
+
+  void SetLevelImpl(LogLevel level);
+  LogLevel GetLevelImpl() const noexcept;
+
+  /// \brief Get the underlying spdlog logger (for advanced usage)
+  std::shared_ptr<void> GetUnderlyingLogger() const { return logger_; }
+
+ private:
+  void LogRawImpl(LogLevel level, const std::string& message) const;
+  void LogWithLocationRawImpl(LogLevel level, const std::string& message,
+                              const std::source_location& location) const;
+
+  std::shared_ptr<void> logger_;  // Type-erased spdlog::logger
+  LogLevel current_level_ = LogLevel::kInfo;
+};
+
+/// \brief No-op logger implementation for performance-critical scenarios
+class NullLogger : public LoggerInterface<NullLogger> {
+ public:
+  bool ShouldLogImpl(LogLevel) const noexcept { return false; }
+
+  template <typename... Args>
+  void LogImpl(LogLevel, std::string_view, Args&&...) const noexcept {}
+
+  template <typename... Args>
+  void LogWithLocationImpl(LogLevel, std::string_view, const std::source_location&,
+                           Args&&...) const noexcept {}
+
+  void SetLevelImpl(LogLevel) noexcept {}
+  LogLevel GetLevelImpl() const noexcept { return LogLevel::kOff; }
+};
+
+/// \brief Global logger registry for managing logger instances
+class LoggerRegistry {
+ public:
+  /// \brief Get the singleton instance of the logger registry
+  static LoggerRegistry& Instance();
+
+  /// \brief Set the default logger implementation
+  ///
+  /// \tparam LoggerImpl The logger implementation type
+  /// \param logger The logger instance to set as default
+  template <typename LoggerImpl>
+  void SetDefaultLogger(std::shared_ptr<LoggerImpl> logger) {
+    static_assert(std::is_base_of_v<LoggerInterface<LoggerImpl>, LoggerImpl>,
+                  "LoggerImpl must inherit from LoggerInterface");
+    default_logger_ = std::static_pointer_cast<void>(logger);
+    log_func_ = [logger](LogLevel level, std::string_view format_str,
+                         const std::string& formatted_message) {
+      if (logger->ShouldLog(level)) {
+        logger->LogImpl(level, "{}", formatted_message);
+      }
+    };
```

Comment:
Ultimately we still do a dynamic call, does it save us anything versus a virtual call? It is significantly more complicated